API/json implementation and interaction with ddev, fixes #477

rfay opened the pull request:
## The Problem/Issue/Bug:

OP #477: It will be done in several sections:

Work to do from #477:

- [x] Add a UserOut logger (just a logrus.New()) 
- [x]  Set a custom formatter like [UserTextFormatter](https://github.com/rfay/logrustry/blob/master/cmd/logrustry/text_formatter.go) to the UserOut - it will drop the timestamp and loglevel, can do other things as well. This is a fork of the logrus TextFormatter.
- [x]  Add a ddev command flag to specify json output, like --json. This would cause setting the UserOut formatter to log.JSONFormatter
- [x] Use logrus (UserOut) for all output aimed at the user
- [x]  Remove all imports of the stdlib log
- [x]  Remove all usage of log in tests, they should be using t.log and family
- [x]  Remove all usages of fmt.Print* and the like
- [x]  `ddev list` and `ddev describe` need more attention than this because they painstakingly build the text as a simple output buffer. We'll need to refactor them so that they build a data structure and then either output the data structure as json or render the table and output it.  
- [x] Retest all interactive prompts to make sure they still look right. For example, database import and files import interactive prompts
- [x] Automated tests for json output

Note that `ddev logs` turns output responsibility over to docker, so can't currently output in json.  We need to consider whether we want to be able to turn container logs into json.

## Design and Structure

This PR changes ddev to use a single output mechanism for all user-destined output. The UserOut log object (based on spf13/logrus) has been modified to output directly to the user, except when the --json flag is provided, in which case it outputs a json response to the same command. So for example, `ddev list` gives traditional output (but using UserOut, not direct stdout), but `ddev list --json` or `ddev list -j` gives the same a machine-readable json version.

## JSON output

For the json `ddev list` and `ddev describe` commands a field named "raw" is added. This is what should be consumed by ddev-ui. The normal msg is there, but it's intended for non-json uses, and would be inappropriate to parse. 

For all other commands, the normal response (or error response) is just wrapped in a json envelope with severity, etc.

## Manual Testing Instructions:

Try all commands. Try the commands with --json (or -j). The json output is pretty-printed when output to a terminal, to make it easier to analyze. 


## Automated Testing Overview:
<!-- Please provide an overview of tests introduced by this PR, or an explanation for why no tests are needed. -->

## Related Issue Link(s):

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

rfay commented:
This still needs tests for the json output, and I'll be doing a walkthrough of the interactive prompts like ddev config, but the basics are in place and it passes tests. I'd really appreciate a review of the code and approach and of the json output. I'm updating the OP to describe what's been done and what it means, and the structure of the json ouput.

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -588,21 +575,28 @@ func (l *LocalApp) Start() error {
 }
 
+// ExecWithTty executes a given command in the container of given type. It allocates a pty for interactive work.
+// Returns ComposeCmd results of err, stdout, stderr
```
this comment line seems inaccurate for this function

rfay replied:
Fixed in 7eb90e2f595dce3c6b1bc27179b1f3fa53428cf3

tannerjfco commented on pkg/plugins/platform/local.go:
```diff
@@ -667,7 +661,7 @@ func (l *LocalApp) DockerEnv() {
 			// @ TODO: I have no idea what a Setenv error would even look like, so I'm not sure what
```
nit, we can probably lose this TODO. I think we're doing all that's reasonable to be done here.

rfay replied:
Hey we've had that in there so long it's an old friend!

tannerjfco commented on pkg/plugins/platform/local_test.go:
```diff
@@ -219,13 +223,11 @@ func TestLocalImportDB(t *testing.T) {
 
-			stdout := testcommon.CaptureStdOut()
-			err = app.Exec("db", true, "mysql", "-e", "SHOW TABLES;")
+			out, _, err := app.Exec("db", "mysql", "-e", "SHOW TABLES;")
```
is the 2nd returned var actually used anywhere we're calling app.Exec? (stderr output i think) I don't think I've seen it in use.

rfay replied:
Yes, the tests now use it instead of trying to capture output in the fancy output-capture way.

tannerjfco commented on pkg/testcommon/testcommon_test.go:
```diff
@@ -57,15 +57,15 @@ func TestChdir(t *testing.T) {
 }
 
-// TestCaptureStdOut ensures capturing of standard out works as expected.
+// TestCaptureStdOut ensures capturing of UserOut works as expected.
```
this is still capture of stdout though no?

rfay replied:
Fixed comment, added minor improvement to capture in the test itself.

tannerjfco commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+		levelColor = nocolor
+	}
+
+	fmt.Fprintf(b, "\x1b[%dm%s\x1b[0m ", levelColor, entry.Message)
```
i'm kind of surprised we have to re-implement handling of so many things like colored output or quoting here. is this also adapted from logrus code?

rfay replied:
Logrus would have been emitting and coloring the log level like [FATA] would be in red, we're coloring the whole thing. That's the change from logrus.

tannerjfco replied:
It'd be helpful to have explanations like this in code comments. I still don't have a firm grasp of what's copied as-is from logrus vs what's been modified, and it'd help to be able to make this determination quickly months later when we've forgotten this discussion.

rfay replied:
I reorganized the formatters (while putting in a separate package for @beeradb ) so that the names are the same as the original files, and there are checked-in versions of the original files in the sirupsen_orig directory. That makes it easy to diff and see what has changed.

tannerjfco commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+		return true
+	}
+	for _, ch := range text {
+		if !((ch >= 'a' && ch <= 'z') ||
```
i'm having a hard time grokking these conditions

rfay replied:
Yeah copied from logrus. No idea.

tannerjfco commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+		if !f.DisableTimestamp {
+			f.appendKeyValue(b, "time", entry.Time.Format(timestampFormat))
+		}
+		//f.appendKeyValue(b, "level", entry.Level.String())
```
nit, cleanup - this and a couple lines later

rfay replied:
Fixed in fa6049a0

tannerjfco commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+// Below is from logus json_formatter.go
```
might be useful to reference a commit/file this is adapted from, and to highlight any of our customizations

rfay replied:
Added in 634f884d

tannerjfco commented on pkg/util/utils.go:
```diff
@@ -1,42 +1,55 @@
 func Failed(format string, a ...interface{}) {
-	Error(format, a...)
-	os.Exit(1)
+	if a != nil {
```
not sure i understand the reason for condition check here

rfay replied:
IIRc, log.Fatalf("some format string", nil) makes ugly output, pointing out the nil to the end user and calling it bad.

tannerjfco commented on pkg/util/utils.go:
```diff
@@ -1,42 +1,55 @@
 func Error(format string, a ...interface{}) {
-	color.Red(format, a...)
+	if a != nil {
+		UserOut.Fatalf(format, a...)
```
fatal doesn't seem right here

rfay replied:
Good catch!

rfay commented on cmd/ddev/cmd/root.go:
```diff
@@ -125,3 +117,7 @@ func dockerNetworkPreRun() {
 }
+
+func init() {
+	RootCmd.PersistentFlags().BoolVarP(&util.JSONOutput, "json", "j", false, "If true, iser-oriented output will be in JSON format.")
```
I'm open to --json-output. -j is in fact nice for testing.

rfay commented:
This now has tests for `ddev list -j`, `ddev describe -j`, and `ddev restart -j`. Mostly "list" and "describe" are special, so deserve tests; restart is just random output, so we're only testing to see if we have some reasonable json.

beeradb commented on pkg/ddevapp/providerPantheon.go:
```diff
@@ -2,11 +2,12 @@ package ddevapp
 	"path/filepath"
 	"strings"
 
+	log "github.com/sirupsen/logrus"
```
There is a `log.Fatalf` in this file which I believe should use the new output logger?

rfay replied:
Fixed in b1367eed8d1ec8475863b7d7b751f1a783dba430

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -80,56 +79,44 @@ func (l *LocalApp) FindContainerByType(containerType string) (docker.APIContaine
-	}
-	appTable := CreateAppTable()
+	shortRoot := RenderHomeRootedDir(l.AppRoot())
+	appDesc := make(map[string]interface{})
```
This would feel a lot better if were a real type rather than `map[string]interface`. The same struct could be used (or we could use a trimmed down version and rely on composition) for the `ddev list` values.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -174,17 +161,17 @@ func (l *LocalApp) ImportDB(imPath string, extPath string) error {
-		fmt.Println("Provide the path to the database you wish to import.")
-		fmt.Println("Import path: ")
+		util.UserOut.Println("Provide the path to the database you wish to import.")
+		fmt.Print("Import path: ")
```
This should be `util.UserOut`, right?

rfay replied:
Unfortunately the logrus system *always* seems to emit a CR, so I used fmt.Print() on the prompts for config and import commands. Open to other solutions.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -174,17 +161,17 @@ func (l *LocalApp) ImportDB(imPath string, extPath string) error {
-			fmt.Println("You provided an archive. Do you want to extract from a specific path in your archive? You may leave this blank if you wish to use the full archive contents")
-			fmt.Println("Archive extraction path:")
+			util.UserOut.Println("You provided an archive. Do you want to extract from a specific path in your archive? You may leave this blank if you wish to use the full archive contents")
+			fmt.Print("Archive extraction path:")
```
This should be `util.UserOut`, right?

rfay replied:
No, same reason, apparently no way to get Logrus to not emit a CR, so prompting looks weird.

beeradb commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// LogSetUp sets up UserOut and log loggers as needed by ddev
+func LogSetUp() {
```
Nit: This feels more like "output" than "logs" to me. I'm not sure why. Maybe there's no difference.

rfay replied:
From the perspective of Logrus, it's log setup. From the perspective of the end user it's output.

beeradb commented on cmd/ddev/cmd/root.go:
```diff
@@ -125,3 +117,7 @@ func dockerNetworkPreRun() {
 }
+
+func init() {
+	RootCmd.PersistentFlags().BoolVarP(&util.JSONOutput, "json-output", "j", false, "If true, user-oriented output will be in JSON format.")
```
It would be nice to leave the door open to future output formats. I'd probably prefer one of the following methods here: 

- Drush style: `--format=json`
- kubectl/helm golang style: `-o/--output=json`

rfay commented on pkg/util/log_support.go:
```diff
@@ -0,0 +1,347 @@
+package util
```
Suggestions for package name?

beeradb replied:
`logger`, `output`, `format`

rfay replied:
There goes Brad wanting a package for every function :)

rfay commented on cmd/ddev/cmd/describe.go:
```diff
@@ -29,25 +30,64 @@ running 'ddev stop <sitename>.`,
-func describeApp(appName string) (string, error) {
-	var err error
+// renderAppDescribe takes the map describing the app and renders it for plain-text output
+func renderAppDescribe(desc map[string]interface{}) (string, error) {
```
I can move this. However, it's only a formatting element, the same as what was here before. It takes the authoritative description (a map) and turns it into a text table (which would only be used by the cobra command). 

As far as the usage of site.Describe() I guess I disagree. It now gives an authoritative machine-usable response (a map), which can be turned into a text table. Having it provide both a textual representation and a map would require extra processing for every call (no big deal) and have 2 returns (odd). But the only place it gets turned into a table is here.

rfay commented on cmd/ddev/cmd/list.go:
```diff
@@ -15,15 +15,25 @@ var DevListCmd = &cobra.Command{
+			platform.RenderAppRow(table, desc)
 		}
+
+		util.UserOut.WithField("raw", appDescs).Print(table)
```
The table is unrelated to the actual json that is output here. There's a kind of a trick going on with this output: The 'raw' field is for the json consumer, and the table is for the plain-text consumer.

rfay commented on cmd/ddev/cmd/list_test.go:
```diff
@@ -3,28 +3,62 @@ package cmd
+
+		// Look through list results in json for this site.
+		found := false
+		for _, listitem := range raw {
```
That's a good suggestion for a thorough test. My opinion is that this is just a one-off test and whatever works is OK here. This is "good enough" for now. It's a slight improvement on what we had.

rfay commented:
I think all the small feedback is accounted for. Now we have decisions to make on larger feedback:

- [ ]  Brad would like to [change the json format flag from a boolean on/off](https://github.com/drud/ddev/pull/499/files#r147212655) to a --format=xxx. I would prefer doing this when it's called for. Right now, the flag is a simple bool and does what is called for in this issue.
- [ ] Brad would [like to use a struct to describe an app](https://github.com/drud/ddev/pull/499/files#r147207529) instead of the more free-form map approach I've taken. This is a valid comment, but the map approach is also valid. I would argue that it doesn't deserve that level of effort or disruption at this time.
- [ ] Brad would like to [move log_support.go to its own package](https://github.com/drud/ddev/pull/499/files#r147216452). I'm fine with it in util. We always disagree about package organization :)
- [ ] Tanner suggested that we might be able to make the util.UserOut.xxx() statements easier by just using log as is, and only having one output format at a time. This does make for simpler lines of code, but it means that we can't use log() separately.  sirupsen logging is a bit of an oddity for our CLI anyway, it's intended for a daemon or long-running process with output, not for something that interacts with the user. However, that's where we are and what we've done for a long time.  So I *could* consolidate everything back into log, it's not that hard. The few remaining log() statements would no longer have timestamps and such, and would be formatted differently.

We should have a quick chat or meeting to resolve these, and I think this will be ready for a final pass when it's green (I seem to have broken something in a few trivial commits).

rickmanelius commented:
Regarding the 4 open items in https://github.com/drud/ddev/pull/499#issuecomment-340852201, for items 1 and 2 I'm all in favor of simplicity for now and expanding when necessary. So if we decide to just keep it a boolean as a toggle and use the existing structure, for now, that probably makes sense unless/until we need the additional effort.

I have no opinion on items 3 and 4.

rfay commented:
There is one more test fix for windows required on this. It affects nothing for review or decisionmaking.

tannerjfco commented on pkg/testcommon/testcommon.go:
```diff
@@ -165,6 +165,34 @@ func Chdir(path string) func() {
+// Capturing starts when it is called. It returns an anonymous function that
+// when called, will return a string containing the output during capture, and
+// revert once again to the original value of os.StdOut.
+func CaptureUserOutOutput() func() string {
```
this name really stutters. CaptureUserOut would be fine IMO

rfay commented:
@tannerjfco Item #4 was based on an earlier discussion in this issue, https://github.com/drud/ddev/pull/499#issuecomment-340852201 - do you want to pursue that at all? We could just use log and lose the "native" capabilities/format of log.Debug() and such. It would make the code less verbose, which would be nice. But then log() is also a kind of not doing log() things at all.

rfay commented:
I think this is ready for maybe a final test. It seems like the important issues are taken care of. 

> - I have to confess, I miss the container output during startup. We don't have any new output until the site is fully ready, and it makes the wait rather uncomfortable.

For this, I made ... output appear to normal users, but the original docker-compose output show if we have DRUD_DEBUG set.  That's an experiment (both ways) to see if you like it, but I like it a little better. I was also missing the time-based feedback.

> - ddev rm and ddev stop give no output until they return successful (or not). Empty newline after the success message on rm.

Fixed

> - "ddev start output was all white text. Previously success messages were colored blue." "Seems like success messages in general are not output as colored text"

This is definitely different than previously, and I was unable to figure out how to make it be the same as before, but I'm not convinced it has to be. The bottom line issue is that we only have one type of output that can be mapped with "success" or "neutral" output, and that's logLevel.info. And we use log-type output for everything that's important. So there was previously an implicit difference (or one-off difference) between things that were explicitly "success" and those that were just informational, but it was never a specific practice. I'm hoping you'll agree to go with this as it is.

> - unexpected newline after ddev exec output

Fixed

> - ddev hostname error outputs as "FATA[0000] Could not add hostname"

Fixed. And thanks especially for this one. It pointed out a very strange setup on the hostname command.

tannerjfco commented:
> > - I have to confess, I miss the container output during startup. We don't have any new output until the site is fully ready, and it makes the wait rather uncomfortable.

> For this, I made ... output appear to normal users, but the original docker-compose output show if we have DRUD_DEBUG set.  That's an experiment (both ways) to see if you like it, but I like it a little better. I was also missing the time-based feedback.

Unfortunately the progress dot output didn't seem to work very well for me. When I ran "ddev start" I got 2 dots, then nothing followed by a bit of wait, and then suddenly the success output emitted. When I ran "ddev rm" no progress dots emitted until the command returned successful.

Overall, I gotta say I really prefer having the compose output. I think its useful to know what its doing, in this case starting containers. On a related note, I also find the "Starting health checks" useful, as we're waiting here for things to report back ok as well. Without this output, we indicate that containers are likely past starting, but no indication of what we're waiting for from there. I'm open to a change in message here, as I know you didn't consider this output useful, but I think its worth saying what we're waiting for.

> > - "ddev start output was all white text. Previously success messages were colored blue." "Seems like success messages in general are not output as colored text"

> This is definitely different than previously, and I was unable to figure out how to make it be the same as before, but I'm not convinced it has to be. The bottom line issue is that we only have one type of output that can be mapped with "success" or "neutral" output, and that's logLevel.info.

I liked having success messages easily distinguished from the rest of the output, but you are right this is not a strict necessity. I'm willing to accept it and move on given the technical limitation involved, and I don't think it should be too disrupting for users.

rfay commented:
So @rmanelius I guess you'll have to make the call. @beeradb  you should chime in. In the API stuff, to prevent uncontrolled output (and supposedly to improve user experience) I replaced all the incomprehensible docker-compose output (Starting this container, etc.) with just dots. DRUD_DEBUG=true brings the old stuff back. My intent was two-fold, 

1. Make all output controlled only by ddev instead of allowing docker-compose to output directly
2.  Eliminate a lot of garbage text that users aren't intended to understand (but that we are comfortable with). 

Now that I went back and built the capability of the dots, it is possible (and already implemented with DRUD_DEBUG=true) to output the stderr from docker-compose (for #1) But I still maintain that our users do not need to see that. 

So decision time.

rickmanelius commented:
@rfay Acknowledged that this is a request for me to weigh in. Attempting to get a test on this today...

rfay commented:
Unblocked me! 

Just so you know @rickmanelius I don't think the color issues here are related to #363. The problem here is that we don't have any way in the logging platform to distinguish between "informational" and "success". And since we're using consistent output techniques, we don't want to be randomly coloring things, but rather use the chosen output to do that. That said, it's probably possible to randomly and manually color a few outputs.

rfay commented:
I re-enabled output of docker-compose output.
And (silly me) I found a simple way of coloring success output in util.Success() of all places. 

There is a newline (sometimes) output in the docker-compose output that I've been unable to track down and work on.
